### PR TITLE
Signals on commit

### PR DESCRIPTION
**Pull request by mkurek**
Multiple fixes in this PR:
* remove `_handle_post_save` variable - previously it was used to mark signals to NOT be executed before M2M relations were saved - the logic which set this field was removed in cbb2486603b0f0c847e4089e7d07ec4fb25bd298, but some code (ex. mail notifications) still relies on it
* fix notifications and publishing dc host update to be done after transaction is committed (and, especially, after M2M relations are saved)

To call some function after committing the transaction, just use:
```
from ralph.signals import post_commit

post_commit(my_handler, MyModel)
```

for example:
```
post_commit(publish_host_update, DataCenterAsset)
```

### CODE REVIEW

File: src/ralph/dns/publishers.py [hunk 1]
```diff
@@ -35,6 +33,7 @@ def _send_once(instance):
     def publish_data_to_dnsaaas(obj):
         return _publish_data_to_dnsaaas(obj)
 
+    # TODO: consider changing it to `ralph.signals.post_commit`
```
Comment (xor-xor) [inline src/ralph/dns/publishers.py]: `# TODO(mkurek):`

File: src/ralph/signals.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+from django.db import connection
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+
+# TODO: make this working as a decorator, example:
```
Comment (xor-xor) [inline src/ralph/signals.py]: `TODO(mkurek):`

File: src/ralph/signals.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            # instance
+            called_attr = '_' + func.__name__ + '_called'
+            if (
+                not getattr(instance, called_attr, False) or
+                not prevent_multiple_calls
+            ):
+                func(instance)
+                setattr(instance, called_attr, True)
+
+        # TODO: replace connection by transaction after upgrading to Django 1.9
```
Comment (xor-xor) [inline src/ralph/signals.py]: `TODO(mkurek):` ;)

File: src/ralph/signals.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+
+# TODO: make this working as a decorator, example:
+# @post_commit(MyModel)
+# def my_handler(instance):
+#    ...
+def post_commit(func, model, signal=post_save, prevent_multiple_calls=True):
+    """
```
Comment (xor-xor) [inline src/ralph/signals.py:11]: +1 for good docstring :)

File: src/ralph/data_center/tests/test_admin.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            )
+        )
+
+    def _prepare_inline_data(self, d):
+        return {
+            '{}{}'.format(self.custom_fields_inline_prefix, k): v
+            for (k, v) in d.items()
+        }
+
+    def test_update_data_center_asset_check_mail_notifications(self):
```
Comment (xor-xor) [inline src/ralph/data_center/tests/test_admin.py]: What about some more descriptive names for tests in this module..? E.g. "test for ... when ...", "test if ... given ..." etc.

**Review by xor-xor — APPROVED** [review]

File: src/ralph/dns/publishers.py [hunk 6]
```diff
@@ -35,6 +33,7 @@ def _send_once(instance):
     def publish_data_to_dnsaaas(obj):
         return _publish_data_to_dnsaaas(obj)
 
+    # TODO(mkurek): consider changing it to `ralph.signals.post_commit`
```
Comment (ghost) [inline src/ralph/dns/publishers.py]: any roadblock to not making it happen now?

File: src/ralph/notifications/sender.py [hunk 7]
```diff
@@ -7,15 +8,24 @@
...
-def send_notification_for_model(sender, instance, **kwargs):
-    if not getattr(instance, '_handle_post_save', False):
-        return
-    ServiceEnvironment = sender.service_env.field.related_model
+def send_notification_for_model(instance):
+    ServiceEnvironment = instance._meta.get_field('service_env').related_model
     old_service_env_id = instance._previous_state['service_env_id']
     new_service_env_id = instance.service_env_id
     if old_service_env_id and old_service_env_id != new_service_env_id:
+        logger.info(
```
Comment (ghost) [inline src/ralph/notifications/sender.py:20]: +1

File: src/ralph/signals.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    @receiver(signal, sender=model, weak=False)
+    def wrap(sender, instance, **kwargs):
+        def wrapper():
+            # prevent from calling the same func multiple times for single
+            # instance
+            called_attr = '_' + func.__name__ + '_called'
+            if (
+                not getattr(instance, called_attr, False) or
+                not prevent_multiple_calls
+            ):
```
Comment (ghost) [inline src/ralph/signals.py:45]: s/called_attr/called_already/ ?
s/prevent_multiple_calls/single_call ?
```
if  not (getattr(instance, called_already, False) and single_call)):
..
```
To me it'd be more readably, what do you think? :)

File: src/ralph/dns/publishers.py (hunk 6 above)
Comment (mkurek) [reply src/ralph/dns/publishers.py]: only time (for testing) :)

**Review by ghost — APPROVED** [review]

File: src/ralph/signals.py (hunk 8 above)
Comment (mkurek) [reply src/ralph/signals.py:45]: I've tuned it slightly. Changed `called_attr` to `called_already_attr` (mix of both approached).

**Review by szok — APPROVED** [review]

**Pull request merged**